packages/esm-patient-test-results-app/src/filter/filter-set.tsx

```diff
@@ -22,43 +23,74 @@ const FilterSet: React.FC<FilterSetProps> = ({ hideFilterSetHeader = false }) =>
+  const [treeDataFiltered, setTreeDataFiltered] = useState(roots);
   const [showSearchInput, setShowSearchInput] = useState(false);
 
+  useEffect(() => {
+    setTreeDataFiltered(roots);
+  }, [roots]);
+
+  function filterTreeNode(inputValue, treeNode) {
+    // If the tree node's display value contains the user input, or any of its children's display contains the user input, return true
+    if (
+      treeNode?.display.toLowerCase().includes(inputValue.toLowerCase()) ||
+      (treeNode?.subSets && treeNode.subSets.some((child) => filterTreeNode(inputValue, child)))
+    ) {
+      return true;
+    }
+
+    // Otherwise, return false
+    return false;
+  }
+
+  function handleInputChange(value) {
+    setSearchTerm(value);
+    const filteredData = roots.filter((node) => filterTreeNode(value, node));
+    setTreeDataFiltered(filteredData);
+  }
+
   return (
     <div className={!tablet ? styles.stickyFilterSet : ''}>
       {!hideFilterSetHeader &&
         (!showSearchInput ? (
           <div className={styles.filterSetHeader}>
             <h4>{t('tree', 'Tree')}</h4>
             <div className={styles.filterSetActions}>
-              <Button
-                kind="ghost"
-                size="sm"
-                onClick={resetTree}
-                renderIcon={(props) => <TreeViewAlt size={16} {...props} />}
-              >
-                {t('resetTreeText', 'Reset tree')}
-              </Button>
+              {searchTerm ? (
+                <Button
+                  kind="ghost"
+                  size="sm"
+                  onClick={resetTree}
+                  renderIcon={(props) => <TreeViewAlt size={16} {...props} />}
+                >
+                  {t('resetTreeText', 'Reset tree')}
+                </Button>
+              ) : null}
+
               <Button kind="ghost" size="sm" renderIcon={SearchIcon} onClick={() => setShowSearchInput(true)}>
                 {t('search', 'Search')}
               </Button>
             </div>
           </div>
         ) : (
           <div className={styles.filterTreeSearchHeader}>
-            <Search size="sm" value={searchTerm} onChange={(evt) => setSearchTerm(evt.target.value)} light />
+            <Search size="sm" value={searchTerm} onChange={(evt) => handleInputChange(evt.target.value)} light />
             <Button kind="secondary" size="sm" onClick={() => {}}>
               {t('search', 'Search')}
             </Button>
             <Button hasIconOnly renderIcon={Close} size="sm" kind="ghost" onClick={() => setShowSearchInput(false)} />
           </div>
         ))}
       <div className={styles.filterSetContent}>
-        {roots?.map((root, index) => (
-          <div className={styles.nestedAccordion} key={`filter-node-${index}`}>
-            <FilterNode root={root} level={0} open={config.concepts[index].defaultOpen} />
-          </div>
-        ))}
+        {treeDataFiltered.length > 0 ? (
+          treeDataFiltered?.map((root, index) => (
```

Comment:
If it's undefined here, the previous line would've thrown...

```suggestion
          treeDataFiltered.map((root, index) => (
```